packages/flutter/lib/src/rendering/object.dart

```diff
@@ -978,19 +1033,31 @@ class PipelineOwner {
         arguments: debugTimelineArguments,
       );
     }
-    assert(() {
-      _debugDoingPaint = true;
-      return true;
-    }());
     try {
+      assert(() {
+        _debugDoingPaint = true;
+        return true;
+      }());
       final List<RenderObject> dirtyNodes = _nodesNeedingPaint;
       _nodesNeedingPaint = <RenderObject>[];
+
       // Sort the dirty nodes in reverse order (deepest first).
       for (final RenderObject node in dirtyNodes..sort((RenderObject a, RenderObject b) => b.depth - a.depth)) {
+        // It is possible that one or more render objects in the paint list are no
+        // longer repaint boundaries. They would have been marked as needing a compositing
+        // bits update, which would mark their parents dirty to the nearest repaint boundary.
+        // Thus it should be safe to drop them from the list.
```

Comment:
Where are the _needsPaint and _needsCompositedLayerUpdate cleared out for these?

Reply:
Can you assert in the if branch (or at the end of painting) that those are not set anymore and the nodes are indeed clean?

Reply:
And if they are no longer dirty, wouldn't it already short-circuit in line 1054 making this code here unnecessary?

Reply:
Its not that they're no longer dirty - they are - its that they aren't repaint boundaries so they can't be a root that we work from for repainting - we need to work from their parents. But if we changed the compositing bits, then we already marked the path to the parent dirty and we'll repaint from it anyway.

The branches on L1054 don't help since they only check if the node is dirty (which it still is)

Reply:
OK, then we should still be able to assert that the nodes we skipped here will be clean at the end of paint (after their parents have repainted), right?

Also - this goes back to one of my previous comments - if we cease to be a repaint boundary, should we just remove the node from the dirty list right then and there?

Reply:
Yeah, we could do that too, I'll try it out